### ISSUE TITLE
Having trouble with mx.vmap and nested mx.grad

### ISSUE THREAD

**Issue by sck-at-ucy** [ml-explore/mlx#1815]
I know I must be doing something stupid but after spending a long time I could not figure it out, so I need some help. 

I am trying to use autograd to compute the laplacian. First derivatives work fine, but when I try to compute second derivatives I get identically zero. I tried removing mx.vmap and doing the 2nd derivatives for a single point, but even then I get zero, so the issue seems to not be related to vmap but rather to how I nest the grads.

```
def T_of_xy(px, py):
    """
    Returns output from the model for one point (px, py).
    """
    out = model(px, py)  # shape: (batch_size, 1), model expects two separate inputs

    return out.squeeze()  # Return the output directly

# Adjusted Laplacian computation
def laplacian_T(model, px, py):
    """
    Compute Laplacian: d^2 T / dx^2 + d^2 T / dy^2 using `grad()`.
    """

    # Compute first derivatives w.r.t. x and y
    dTdx = mx.vmap(mx.grad(T_of_xy, argnums=0))(px, py)  # Gradient w.r.t. px
    dTdy = mx.vmap(mx.grad(T_of_xy, argnums=1))(px, py)  # Gradient w.r.t. py

    # Check the first derivatives (debugging step)
    print(f'dTdx: {dTdx}')
    print(f'dTdy: {dTdy}')


    # Compute second derivatives w.r.t. x and y
    d2Tdx2 = mx.vmap(mx.grad(mx.grad(T_of_xy, argnums=0), argnums=0))(px, py)  # Second derivative w.r.t. px
    d2Tdy2 = mx.vmap(mx.grad(mx.grad(T_of_xy, argnums=1), argnums=1))(px, py)  # Second derivative w.r.t. py

    # Check the second derivatives (debugging step)
    print(f'd2Tdx2: {d2Tdx2}')
    print(f'd2Tdy2: {d2Tdy2}')

    # Compute Laplacian (sum of second derivatives)
    laplacian = d2Tdx2 + d2Tdy2

    # Debugging the Laplacian
    print(f'Laplacian: {laplacian}')

    return mx.array(laplacian)
```

The laplacian remains identically zero

dTdx: array([-0.000360303, -0.000290909, -0.000186427, ..., -0.000386215, -0.000519069, -0.000334836], dtype=float32)
dTdy: array([-6.11922e-05, -0.000122368, -0.000250414, ..., -8.05613e-05, -0.000204918, -3.1627e-05], dtype=float32)
d2Tdx2: array([0, 0, 0, ..., 0, 0, 0], dtype=float32)
d2Tdy2: array([0, 0, 0, ..., 0, 0, 0], dtype=float32) 
Laplacian: array([0, 0, 0, ..., 0, 0, 0], dtype=float32)

**Comment by angeloskath** [comment]
I think it depends on the model you are using. Could it be that it doesn't have 2nd order gradients? For instance relu networks do not have 2nd order gradients. They are not linear functions but they are piecewise linear.

The following toy examples makes it clear I believe

```python
import mlx.core as mx
import mlx.nn as nn


class Model(nn.Module):
    def __init__(self, activation="relu"):
        super().__init__()
        self.activation = nn.relu if activation == "relu" else mx.tanh
        self.linears = [
            nn.Linear(2, 10),
            nn.Linear(10, 10),
            nn.Linear(10, 10),
        ]
        self.final = nn.Linear(10, 1)

    def __call__(self, px, py):
        p = mx.stack([px, py], axis=-1)
        for l in self.linears:
            p = self.activation(l(p))
        return self.final(p)


# model = Model(activation="relu")
model = Model(activation="tanh")


def T_of_xy(px, py):
    out = model(px, py)
    return out.squeeze()


px = mx.random.uniform(shape=(10,))
py = mx.random.uniform(shape=(10,))
dTdx = mx.vmap(mx.grad(T_of_xy, argnums=0))(px, py)
dTdy = mx.vmap(mx.grad(T_of_xy, argnums=1))(px, py)
d2Tdx2 = mx.vmap(mx.grad(mx.grad(T_of_xy, argnums=0), argnums=0))(px, py)
d2Tdy2 = mx.vmap(mx.grad(mx.grad(T_of_xy, argnums=1), argnums=1))(px, py)
print(d2Tdx2)
print(d2Tdy2)

```

**Comment by sck-at-ucy** [comment]
Thank you. Yes indeed, this is exactly my issue, I overlooked the piecewise linearity of the MLP with ReLU
